keylime/mba/mba.py

```diff
@@ -21,46 +46,98 @@
 MBAgg = Optional[Dict[str, List[str]]]
 
 # ###########
-# import/load policy engine
+# list of all imported modules for MBA.
 # ###########
 
+_mba_imports = []
 
-def load_policy_engine() -> None:
+# ###########
+# import/load measured boot attestation imports
+# ###########
+
+
+def load_imports() -> None:
     """
-    MBA API front-end for importing the policy engine.
-    No inputs, but implementation *will* read from keylime configuration files.
+    MBA API front-end for importing any modules needed by measured boot attestation.
+    No inputs.
     No outputs.
-    Side effects: policy engine is now loaded.
-    Exceptions: If the policy engine cannot be loaded.
+    Side effects: Reads from keylime configuration files.
+      After execution all imports required by MBA are loaded and initialized.
+    Exceptions: If any policy engine in the list cannot be loaded.
+    """
+    try:
+        imports = config.getlist("verifier", "measured_boot_imports")
+        # these are the defaults
+        imports.append("keylime.mba.elchecking.elchecker")
+        imports.append("keylime.mba.elchecking.example")
+        imports.append("keylime.mba.elparsing.tpm2_tools_elparser")
+        # initialize all modules if they carry an initialization function.
+        for m in imports:
+            _mba_imports.append(importlib.import_module(m, __package__))
+    except Exception as e:
+        raise ValueError from e
+
+
+# ##########
+# parse a boot log to JSON
+# ##########
+
+
+def bootlog_parse(bootlog_b64: Optional[str], hash_alg: str) -> Tuple[MBPCRDict, MBAgg, MBLog, Failure]:
+    """
+    MBA API front-end for parsing a binary boot log.
+
+    :param bootlog_b64: a b64 encoded version of the binary boot eventlog from the agent's /sys/kernel directory
+    :param hash_alg: the expected hash algorithm to use when decoding the boot log. This is a string, something like
+    "sha1" or "sha256".
+    :returns:
+    * mb_pcrs_hashes: the expected hash values for all PCRs mentioned in the boot log, using the input hash_alg.
+    * boot_aggregates: all boot aggregates calculated by the parsed event log.
+      This typically includes the aggregate of PCRs 0 to 9, although for sha1 algorithms we only consider PCRs 0 to 7.
+    * mb_data: the actual decoded boot log, as a JSON array of events.
+    * mb_failure: a list of all failures encountered while parsing the event log.
     """
+
     try:
-        policies.load_policies()
+        m = _find_implementation("bootlog_parse")
+        return m.bootlog_parse(bootlog_b64, hash_alg)  # type: ignore[no-any-return]
     except Exception as e:
         raise ValueError from e
 
 
 # ##########
 # read a policy definition (aka "refstate") from a file.
+# TODO: do we need a policy file validator?
```

Comment:
is not the task of policy_is_valid defined later?

Reply:
Yes. Documentation fail. Fixing.